include/oxen/quic/endpoint.hpp

```diff
@@ -381,11 +376,42 @@ namespace oxen::quic
         void send_or_queue_packet(
                 const Path& p, std::vector<std::byte> buf, uint8_t ecn, std::function<void(io_result)> callback = nullptr);
 
+        void send_stateless_reset(const Packet& pkt, quic_cid& cid);
+
+        Connection* check_stateless_reset(const Packet& pkt, quic_cid& cid);
+
         void send_version_negotiation(const ngtcp2_version_cid& vid, Path p);
 
         void check_timeouts();
 
-        Connection* accept_initial_connection(const Packet& pkt);
+        Connection* accept_initial_connection(const Packet& pkt, quic_cid& cid);
+
+        template <typename... Opt>
+        static constexpr void check_for_tls_creds()
+        {
+            static_assert(
+                    (0 + ... + std::is_convertible_v<std::remove_cvref_t<Opt>, std::shared_ptr<TLSCreds>>) == 1,
+                    "Endpoint listen/connect require exactly one std::shared_ptr<TLSCreds> argument");
+        }
+
+        template <typename... Opt>
+        static void check_verification_scheme(Endpoint* e)
+        {
+            if constexpr ((std::is_same_v<opt::disable_key_verification, std::remove_cvref_t<Opt>> || ...))
+            {
+                if (e->zero_rtt_enabled())
+                    throw std::invalid_argument{"Disabling key verification is incompatible with 0rtt ticketing!"};
+            }
+        }
+
+        template <concepts::quic_address_type T, typename... Opt>
+        static constexpr void check_address_scheme()
+        {
+            if constexpr ((std::is_same_v<opt::disable_key_verification, std::remove_cvref_t<Opt>> || ...))
+            {
+                static_assert(std::is_same_v<T, Address>, "Disabling key verification requires keyless address!");
+            }
```

Comment:
This makes this fail to compile, which seems good:
```C++
    RemoteAddress a = /*...*/;
    ep->connect(a, opt::disable_key_verification{});
```
but shouldn't it also check that if I use a plain Address, I *am* explicitly specifying `opt::disable_key_verification{}`, so that this is not allowed:
```C++
    Address a = /*...*/;
    ep->connect(a);
```
because I'm pretty sure this will fail, because it will not have `disable_key_verification = true` applied to the context.

That also makes me think that the `opt` here isn't strictly needed because you can always infer it based on whether the address is an `Address` or `RemoteAddress`.  That said: perhaps it's still worth requiring the explicit option to actively disable this, given the implications of the option.

Reply:
Good call on the check on `RemoteAddress`. As to the opt, I agree that it's worth requiring the explicit option; this way we keep a logical framework in place to modify/add any functionality to our implementation